gse-afs-ext-base/src/main/java/com/powsybl/gse/afs/ext/base/ContingencyStoreEditor.java

```diff
@@ -45,29 +46,88 @@
+            if (getTreeItem() != null) {
+                if (getTreeItem().getValue() instanceof Contingency) {
+                    contingencyItem = getTreeItem();
+                } else if (getTreeItem().getValue() instanceof ContingencyElement) {
+                    contingencyItem = getTreeItem().getParent();
+                }
+            }
+            return contingencyItem;
+        }
+
+        private EquipmentInfo createEquipmentInfo(ContingencyElement contingencyElement) {
+            ContingencyElementType type = contingencyElement.getType();
+            if (type.equals(ContingencyElementType.BUSBAR_SECTION)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "BUSBAR_SECTION");
+            } else if (type.equals(ContingencyElementType.GENERATOR)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "GENERATOR");
+            } else if (type.equals(ContingencyElementType.HVDC_LINE)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "HVDC_LINE");
+            } else if (type.equals(ContingencyElementType.BRANCH)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "BRANCH");
+            } else {
+                return null;
+            }
+        }
+
+        private void onDragDetected(MouseEvent event) {
+            TreeItem<Object> item = contingencyTree.getSelectionModel().getSelectedItem();
+            if (item.getValue() instanceof ContingencyElement) {
+                Dragboard db = startDragAndDrop(TransferMode.COPY_OR_MOVE);
+                ClipboardContent content = new ClipboardContent();
+                content.put(EquipmentInfo.DATA_FORMAT, createEquipmentInfo((ContingencyElement) item.getValue()));
+                db.setContent(content);
+                event.consume();
+            }
+        }
+
+        private void onDragOver(DragEvent event) {
+            Dragboard db = event.getDragboard();
+            if (!(event.getGestureSource() instanceof ContingencyTreeCell) ||
+                    checkTreeItem() != ((ContingencyTreeCell) event.getGestureSource()).checkTreeItem()) {
                 if (db.hasContent(EquipmentInfo.DATA_FORMAT)) {
-                    EquipmentInfo equipmentInfo = (EquipmentInfo) db.getContent(EquipmentInfo.DATA_FORMAT);
+                    acceptSingleDraggedElement(event);
+                } else if (db.hasContent(EquipmentInfo.DATA_FORMAT_LIST)) {
+                    acceptMultipleDraggedElements(event);
+                }
+            }
+            event.consume();
+        }
+
+        private void onDragDropped(DragEvent event) {
+            if (event.getTransferMode().equals(TransferMode.MOVE) &&
+                    event.getGestureSource() instanceof ContingencyTreeCell) {
+                remove();
```

Comment:
If something goes wrong later, you can lost a Contingency item by removing it too early. I think you've to remove the item only if `success = true`, at the end of this method.